1.15.4: pytest is failing

- [x] I have searched the issue tracker and believe that this is not a duplicate.
I'm trying to package your module as an rpm package. So I'm using the typical PEP517 based build, install and test cycle used on building packages from non-root account.
- `python3 -sBm build -w --no-isolation`
- because I'm calling `build` with `--no-isolation` I'm using during all processes only locally installed modules
- install .whl file in </install/prefix>
- run pytest with PYTHONPATH pointing to sitearch and sitelib inside </install/prefix>

Here is pytest output:
<details>

```console
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-pdm-1.15.4-2.fc35.x86_64/usr/lib64/python3.8/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-pdm-1.15.4-2.fc35.x86_64/usr/lib/python3.8/site-packages
+ /usr/bin/pytest -ra
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/tkloczko/rpmbuild/BUILD/pdm-1.15.4, configfile: pyproject.toml
plugins: mock-3.8.1
collected 408 items

tests/test_formats.py ...........                                                                                                                                    [  2%]
tests/test_installer.py ...........                                                                                                                                  [  5%]
tests/test_integration.py FFF.F.                                                                                                                                     [  6%]
tests/test_plugin.py ....                                                                                                                                            [  7%]
tests/test_project.py .............F                                                                                                                                 [ 11%]
tests/test_signals.py ..                                                                                                                                             [ 11%]
tests/test_utils.py ...............................                                                                                                                  [ 19%]
tests/cli/test_add.py .........................                                                                                                                      [ 25%]
tests/cli/test_build.py .............                                                                                                                                [ 28%]
tests/cli/test_cache.py ............                                                                                                                                 [ 31%]
tests/cli/test_config.py .........                                                                                                                                   [ 33%]
tests/cli/test_install.py .................                                                                                                                          [ 37%]
tests/cli/test_list.py ............                                                                                                                                  [ 40%]
tests/cli/test_lock.py .....                                                                                                                                         [ 42%]
tests/cli/test_others.py .....................F.                                                                                                                     [ 47%]
tests/cli/test_plugin.py ...                                                                                                                                         [ 48%]
tests/cli/test_remove.py ..........                                                                                                                                  [ 50%]
tests/cli/test_run.py .............F........                                                                                                                         [ 56%]
tests/cli/test_update.py .............                                                                                                                               [ 59%]
tests/cli/test_use.py .....                                                                                                                                          [ 60%]
tests/models/test_candidates.py .............................                                                                                                        [ 67%]
tests/models/test_marker.py ........                                                                                                                                 [ 69%]
tests/models/test_requirements.py ..................                                                                                                                 [ 74%]
tests/models/test_setup_parsing.py ..........                                                                                                                        [ 76%]
tests/models/test_specifiers.py ..............................................                                                                                       [ 87%]
tests/models/test_versions.py ...................                                                                                                                    [ 92%]
tests/resolver/test_resolve.py ..............................                                                                                                        [100%]

================================================================================= FAILURES =================================================================================
_______________________________________________________________________ test_basic_integration[2.7] ________________________________________________________________________

python_version = '2.7', core = <pdm.core.Core object at 0x7f3b6578efa0>, tmp_path = PosixPath('/tmp/pytest-of-tkloczko/pytest-177/test_basic_integration_2_7_0')
invoke = <function invoke.<locals>.caller at 0x7f3b662805e0>

    @pytest.mark.integration
    @pytest.mark.network
    @pytest.mark.parametrize("python_version", ["2.7", "3.6", "3.7", "3.8", "3.9"])
    def test_basic_integration(python_version, core, tmp_path, invoke):
        """An e2e test case to ensure PDM works on all supported Python versions"""
        project = core.create_project(tmp_path)
        project.root.joinpath("foo.py").write_text("import django\n")
        additional_args = ["--no-self"] if python_version == "2.7" else []
>       invoke(["use", "-f", python_version], obj=project, strict=True)

tests/test_integration.py:14:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = ['use', '-f', '2.7'], strict = True, kwargs = {'obj': <Project '/tmp/pytest-of-tkloczko/pytest-177/test_basic_integration_2_7_0'>}, result = <Result SystemExit(1)>

    def caller(args, strict=False, **kwargs):
        result = runner.invoke(
            core, args, catch_exceptions=not strict, prog_name="pdm", **kwargs
        )
        if strict and result.exit_code != 0:
>           raise RuntimeError(
                f"Call command {args} failed({result.exit_code}): {result.stderr}"
            )
E           RuntimeError: Call command ['use', '-f', '2.7'] failed(1): [NoPythonVersion]: Python interpreter is not found on the system.
E           Add '-v' to see the detailed traceback

tests/conftest.py:359: RuntimeError
_______________________________________________________________________ test_basic_integration[3.6] ________________________________________________________________________

python_version = '3.6', core = <pdm.core.Core object at 0x7f3b6574c1f0>, tmp_path = PosixPath('/tmp/pytest-of-tkloczko/pytest-177/test_basic_integration_3_6_0')
invoke = <function invoke.<locals>.caller at 0x7f3b6573eca0>

    @pytest.mark.integration
    @pytest.mark.network
    @pytest.mark.parametrize("python_version", ["2.7", "3.6", "3.7", "3.8", "3.9"])
    def test_basic_integration(python_version, core, tmp_path, invoke):
        """An e2e test case to ensure PDM works on all supported Python versions"""
        project = core.create_project(tmp_path)
        project.root.joinpath("foo.py").write_text("import django\n")
        additional_args = ["--no-self"] if python_version == "2.7" else []
>       invoke(["use", "-f", python_version], obj=project, strict=True)

tests/test_integration.py:14:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = ['use', '-f', '3.6'], strict = True, kwargs = {'obj': <Project '/tmp/pytest-of-tkloczko/pytest-177/test_basic_integration_3_6_0'>}, result = <Result SystemExit(1)>

    def caller(args, strict=False, **kwargs):
        result = runner.invoke(
            core, args, catch_exceptions=not strict, prog_name="pdm", **kwargs
        )
        if strict and result.exit_code != 0:
>           raise RuntimeError(
                f"Call command {args} failed({result.exit_code}): {result.stderr}"
            )
E           RuntimeError: Call command ['use', '-f', '3.6'] failed(1): [NoPythonVersion]: Python interpreter is not found on the system.
E           Add '-v' to see the detailed traceback

tests/conftest.py:359: RuntimeError
_______________________________________________________________________ test_basic_integration[3.7] ________________________________________________________________________

python_version = '3.7', core = <pdm.core.Core object at 0x7f3b661cac70>, tmp_path = PosixPath('/tmp/pytest-of-tkloczko/pytest-177/test_basic_integration_3_7_0')
invoke = <function invoke.<locals>.caller at 0x7f3b655161f0>

    @pytest.mark.integration
    @pytest.mark.network
    @pytest.mark.parametrize("python_version", ["2.7", "3.6", "3.7", "3.8", "3.9"])
    def test_basic_integration(python_version, core, tmp_path, invoke):
        """An e2e test case to ensure PDM works on all supported Python versions"""
        project = core.create_project(tmp_path)
        project.root.joinpath("foo.py").write_text("import django\n")
        additional_args = ["--no-self"] if python_version == "2.7" else []
>       invoke(["use", "-f", python_version], obj=project, strict=True)

tests/test_integration.py:14:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = ['use', '-f', '3.7'], strict = True, kwargs = {'obj': <Project '/tmp/pytest-of-tkloczko/pytest-177/test_basic_integration_3_7_0'>}, result = <Result SystemExit(1)>

    def caller(args, strict=False, **kwargs):
        result = runner.invoke(
            core, args, catch_exceptions=not strict, prog_name="pdm", **kwargs
        )
        if strict and result.exit_code != 0:
>           raise RuntimeError(
                f"Call command {args} failed({result.exit_code}): {result.stderr}"
            )
E           RuntimeError: Call command ['use', '-f', '3.7'] failed(1): [NoPythonVersion]: Python interpreter is not found on the system.
E           Add '-v' to see the detailed traceback

tests/conftest.py:359: RuntimeError
_______________________________________________________________________ test_basic_integration[3.9] ________________________________________________________________________

python_version = '3.9', core = <pdm.core.Core object at 0x7f3b65515a60>, tmp_path = PosixPath('/tmp/pytest-of-tkloczko/pytest-177/test_basic_integration_3_9_0')
invoke = <function invoke.<locals>.caller at 0x7f3b64731e50>

    @pytest.mark.integration
    @pytest.mark.network
    @pytest.mark.parametrize("python_version", ["2.7", "3.6", "3.7", "3.8", "3.9"])
    def test_basic_integration(python_version, core, tmp_path, invoke):
        """An e2e test case to ensure PDM works on all supported Python versions"""
        project = core.create_project(tmp_path)
        project.root.joinpath("foo.py").write_text("import django\n")
        additional_args = ["--no-self"] if python_version == "2.7" else []
>       invoke(["use", "-f", python_version], obj=project, strict=True)

tests/test_integration.py:14:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = ['use', '-f', '3.9'], strict = True, kwargs = {'obj': <Project '/tmp/pytest-of-tkloczko/pytest-177/test_basic_integration_3_9_0'>}, result = <Result SystemExit(1)>

    def caller(args, strict=False, **kwargs):
        result = runner.invoke(
            core, args, catch_exceptions=not strict, prog_name="pdm", **kwargs
        )
        if strict and result.exit_code != 0:
>           raise RuntimeError(
                f"Call command {args} failed({result.exit_code}): {result.stderr}"
            )
E           RuntimeError: Call command ['use', '-f', '3.9'] failed(1): [NoPythonVersion]: Python interpreter is not found on the system.
E           Add '-v' to see the detailed traceback

tests/conftest.py:359: RuntimeError
______________________________________________________________________ test_set_non_exist_python_path ______________________________________________________________________

project_no_init = <Project '/tmp/pytest-of-tkloczko/pytest-177/test_set_non_exist_python_path0'>

    @pytest.mark.path
    def test_set_non_exist_python_path(project_no_init):
        project_no_init.project_config["python.path"] = "non-exist-python"
        project_no_init._python = None
>       assert project_no_init.python.executable == Path(sys.executable)
E       AssertionError: assert PosixPath('/usr/bin/python3.8') == PosixPath('/usr/bin/python3')
E        +  where PosixPath('/usr/bin/python3.8') = <pdm.models.python.PythonInfo object at 0x7f3b656af2e0>.executable
E        +    where <pdm.models.python.PythonInfo object at 0x7f3b656af2e0> = <Project '/tmp/pytest-of-tkloczko/pytest-177/test_set_non_exist_python_path0'>.python
E        +  and   PosixPath('/usr/bin/python3') = Path('/usr/bin/python3')
E        +    where '/usr/bin/python3' = sys.executable

tests/test_project.py:216: AssertionError
-------------------------------------------------------------------------- Captured stdout setup ---------------------------------------------------------------------------
Using Python interpreter: /usr/bin/python3.8 (3.8)
_________________________________________________________________________ test_completion_command __________________________________________________________________________

invoke = <function invoke.<locals>.caller at 0x7f3b5df17d30>

    def test_completion_command(invoke):
        result = invoke(["completion", "bash"])
>       assert result.exit_code == 0
E       assert 1 == 0
E        +  where 1 = <Result SystemExit(1)>.exit_code

tests/cli/test_others.py:244: AssertionError
_____________________________________________________________________ test_run_script_with_dotenv_file _____________________________________________________________________

project = <Project '/tmp/pytest-of-tkloczko/pytest-177/test_run_script_with_dotenv_fi0'>, invoke = <function invoke.<locals>.caller at 0x7f3b5dd7ac10>
capfd = <_pytest.capture.CaptureFixture object at 0x7f3b5d99dfd0>

    def test_run_script_with_dotenv_file(project, invoke, capfd):
        (project.root / "test_script.py").write_text("import os; print(os.getenv('FOO'))")
        project.tool_settings["scripts"] = {
            "test_script": {"cmd": "python test_script.py", "env_file": ".env"}
        }
        project.write_pyproject()
        (project.root / ".env").write_text("FOO=bar")
        capfd.readouterr()
        with cd(project.root):
            invoke(["run", "test_script"], obj=project)
>           assert capfd.readouterr()[0].strip() == "bar"
E           AssertionError: assert '' == 'bar'
E             - bar

tests/cli/test_run.py:207: AssertionError
-------------------------------------------------------------------------- Captured stdout setup ---------------------------------------------------------------------------
Using Python interpreter: /usr/bin/python3.8 (3.8)
Changes are written to pyproject.toml.
========================================================================= short test summary info ==========================================================================
FAILED tests/test_integration.py::test_basic_integration[2.7] - RuntimeError: Call command ['use', '-f', '2.7'] failed(1): [NoPythonVersion]: Python interpreter is not f...
FAILED tests/test_integration.py::test_basic_integration[3.6] - RuntimeError: Call command ['use', '-f', '3.6'] failed(1): [NoPythonVersion]: Python interpreter is not f...
FAILED tests/test_integration.py::test_basic_integration[3.7] - RuntimeError: Call command ['use', '-f', '3.7'] failed(1): [NoPythonVersion]: Python interpreter is not f...
FAILED tests/test_integration.py::test_basic_integration[3.9] - RuntimeError: Call command ['use', '-f', '3.9'] failed(1): [NoPythonVersion]: Python interpreter is not f...
FAILED tests/test_project.py::test_set_non_exist_python_path - AssertionError: assert PosixPath('/usr/bin/python3.8') == PosixPath('/usr/bin/python3')
FAILED tests/cli/test_others.py::test_completion_command - assert 1 == 0
FAILED tests/cli/test_run.py::test_run_script_with_dotenv_file - AssertionError: assert '' == 'bar'
================================================================ 7 failed, 401 passed in 284.42s (0:04:44) =================================================================
```
</details>

Okay, after devendoring the calls to `pip._vendor` I am down to 5 failing tests! :tada: 

And those are solely the ones that should be skipped if the specific interpreter is not found:

```
=========================== short test summary info ============================
FAILED ../../../../dev/tests/test_integration.py::test_basic_integration[2.7]
FAILED ../../../../dev/tests/test_integration.py::test_basic_integration[3.6]
FAILED ../../../../dev/tests/test_integration.py::test_basic_integration[3.7]
FAILED ../../../../dev/tests/test_integration.py::test_basic_integration[3.8]
FAILED ../../../../dev/tests/test_integration.py::test_basic_integration[3.9]
=========== 5 failed, 403 passed, 597 warnings in 138.60s (0:02:18) ============
```

--- a/news/1175.bugfix.md
+++ b/news/1175.bugfix.md
@@ -1,0 +1,1 @@
+Fix the object identity issue due to inconsistent import paths of pip._vendor package.

--- a/pdm/__init__.py
+++ b/pdm/__init__.py
@@ -16,14 +16,3 @@ __all__ = (
     "Synchronizer",
     "Core",
 )
-
-
-def _fix_pkg_resources() -> None:
-    import importlib
-    import sys
-
-    sys.modules["pkg_resources"] = importlib.import_module("pip._vendor.pkg_resources")
-
-
-_fix_pkg_resources()
-del _fix_pkg_resources

--- a/pdm/cli/actions.py
+++ b/pdm/cli/actions.py
@@ -826,7 +826,7 @@ def check_update(project: Project) -> None:
     import sys
     from shlex import quote
 
-    from pip._vendor.packaging.version import parse as parse_version
+    from packaging.version import parse as parse_version
 
     from pdm.cli.utils import (
         is_homebrew_installation,

--- a/pdm/models/markers.py
+++ b/pdm/models/markers.py
@@ -4,13 +4,15 @@ import operator
 from functools import reduce
 from typing import Any, List, Optional, Set, Tuple, Union
 
-from pip._vendor.packaging.markers import Marker as PackageMarker
+from packaging.markers import Marker as _Marker
 
 from pdm.models.specifiers import PySpecSet
-from pdm.utils import join_list_with
+from pdm.utils import import_pip_vendor_object, join_list_with
+
+MARKER_TYPES = import_pip_vendor_object("packaging.markers", "Marker")
 
 
-class Marker(PackageMarker):
+class Marker(_Marker):
     """A subclass of Marker that supports union and intersection merging."""
 
     def copy(self) -> "Marker":
@@ -19,7 +21,7 @@ class Marker(PackageMarker):
         return inst
 
     def __eq__(self, other: Any) -> bool:
-        if not isinstance(other, PackageMarker):
+        if not isinstance(other, MARKER_TYPES):
             return False
         return str(self) == str(other)
 
@@ -48,7 +50,7 @@ class Marker(PackageMarker):
         return marker, _build_pyspec_from_marker(join_list_with(py_markers, "and"))
 
 
-def get_marker(marker: Union[PackageMarker, Marker, None]) -> Optional[Marker]:
+def get_marker(marker: Optional[_Marker]) -> Optional[Marker]:
     return Marker(str(marker)) if marker else None
 
 
@@ -79,8 +81,8 @@ def split_marker_extras(marker: str) -> Tuple[Set[str], str]:
 
     if not marker:
         return set(), marker
-    new_marker = PackageMarker(marker)
-    submarkers = PackageMarker(marker)._markers
+    new_marker = _Marker(marker)
+    submarkers = _Marker(marker)._markers
     if "or" in submarkers:
         extras = extract_extras(submarkers)
         if extras:

--- a/pdm/models/requirements.py
+++ b/pdm/models/requirements.py
@@ -12,11 +12,9 @@ import warnings
 from pathlib import Path
 from typing import Any, Sequence, Type, TypeVar, cast
 
-from pip._vendor.packaging.markers import InvalidMarker
-from pip._vendor.packaging.requirements import InvalidRequirement
-from pip._vendor.packaging.specifiers import SpecifierSet
+from packaging.specifiers import SpecifierSet
 from pip._vendor.pkg_resources import Requirement as PackageRequirement
-from pip._vendor.pkg_resources import RequirementParseError, safe_name
+from pip._vendor.pkg_resources import safe_name
 
 from pdm._types import RequirementDict
 from pdm.exceptions import ExtrasWarning, RequirementError
@@ -33,6 +31,7 @@ from pdm.models.setup import Setup
 from pdm.models.specifiers import PySpecSet, get_specifier
 from pdm.utils import (
     add_ssh_scheme_to_git_uri,
+    import_pip_vendor_object,
     parse_name_version_from_wheel,
     url_without_fragments,
 )
@@ -41,6 +40,12 @@ if sys.version_info >= (3, 8):
     from importlib.metadata import Distribution
 else:
     from importlib_metadata import Distribution
+
+INVAID_MARKER_ERROR = import_pip_vendor_object("packaging.markers", "InvalidMarker")
+INVALID_REQUIREMENT_ERROR = (
+    *import_pip_vendor_object("packaging.requirements", "InvalidRequirement"),
+    *import_pip_vendor_object("pkg_resources", "RequirementParseError"),
+)
 
 VCS_SCHEMA = ("git", "hg", "svn", "bzr")
 _vcs_req_re = re.compile(
@@ -148,7 +153,7 @@ class Requirement:
         if "marker" in kwargs:
             try:
                 kwargs["marker"] = get_marker(kwargs["marker"])
-            except InvalidMarker as e:
+            except INVAID_MARKER_ERROR as e:
                 raise RequirementError("Invalid marker: %s" % str(e)) from None
         if "extras" in kwargs and isinstance(kwargs["extras"], str):
             kwargs["extras"] = tuple(
@@ -487,7 +492,7 @@ def parse_requirement(line: str, editable: bool = False) -> Requirement:
     else:
         try:
             package_req = PackageRequirement(line)  # type: ignore
-        except (RequirementParseError, InvalidRequirement) as e:
+        except INVALID_REQUIREMENT_ERROR as e:
             m = _file_req_re.match(line)
             if m is None:
                 raise RequirementError(str(e)) from None

--- a/pdm/models/requirements.py
+++ b/pdm/models/requirements.py
@@ -41,7 +41,7 @@ if sys.version_info >= (3, 8):
 else:
     from importlib_metadata import Distribution
 
-INVAID_MARKER_ERROR = import_pip_vendor_object("packaging.markers", "InvalidMarker")
+INVALID_MARKER_ERROR = import_pip_vendor_object("packaging.markers", "InvalidMarker")
 INVALID_REQUIREMENT_ERROR = (
     *import_pip_vendor_object("packaging.requirements", "InvalidRequirement"),
     *import_pip_vendor_object("pkg_resources", "RequirementParseError"),
@@ -153,7 +153,7 @@ class Requirement:
         if "marker" in kwargs:
             try:
                 kwargs["marker"] = get_marker(kwargs["marker"])
-            except INVAID_MARKER_ERROR as e:
+            except INVALID_MARKER_ERROR as e:
                 raise RequirementError("Invalid marker: %s" % str(e)) from None
         if "extras" in kwargs and isinstance(kwargs["extras"], str):
             kwargs["extras"] = tuple(

--- a/pdm/models/requirements.py
+++ b/pdm/models/requirements.py
@@ -14,7 +14,7 @@ from typing import Any, Sequence, Type, TypeVar, cast
 
 from packaging.specifiers import SpecifierSet
 from pip._vendor.pkg_resources import Requirement as PackageRequirement
-from pip._vendor.pkg_resources import safe_name
+from pip._vendor.pkg_resources import RequirementParseError, safe_name
 
 from pdm._types import RequirementDict
 from pdm.exceptions import ExtrasWarning, RequirementError
@@ -43,8 +43,8 @@ else:
 
 INVALID_MARKER_ERROR = import_pip_vendor_object("packaging.markers", "InvalidMarker")
 INVALID_REQUIREMENT_ERROR = (
+    RequirementParseError,
     *import_pip_vendor_object("packaging.requirements", "InvalidRequirement"),
-    *import_pip_vendor_object("pkg_resources", "RequirementParseError"),
 )
 
 VCS_SCHEMA = ("git", "hg", "svn", "bzr")

--- a/pdm/models/specifiers.py
+++ b/pdm/models/specifiers.py
@@ -6,18 +6,20 @@ from operator import attrgetter
 from pathlib import Path
 from typing import Any, Iterable, List, Set, Tuple, Union, cast
 
-from pip._vendor.packaging.specifiers import SpecifierSet
+from packaging.specifiers import SpecifierSet
 
 from pdm.exceptions import InvalidPyVersion
 from pdm.models.versions import Version
+from pdm.utils import import_pip_vendor_object
 
 MAX_VERSIONS_FILE = Path(__file__).with_name("python_max_versions.json")
+SPECIFIER_TYPES = import_pip_vendor_object("packaging.specifiers", "SpecifierSet")
 
 
 @lru_cache()
 def get_specifier(version_str: Union[SpecifierSet, str]) -> SpecifierSet:
-    if isinstance(version_str, SpecifierSet):
-        return version_str
+    if isinstance(version_str, SPECIFIER_TYPES):
+        return cast(SpecifierSet, version_str)
     if not version_str or version_str == "*":
         return SpecifierSet()
     return SpecifierSet(version_str)

--- a/pdm/utils.py
+++ b/pdm/utils.py
@@ -32,8 +32,8 @@ from typing import (
     overload,
 )
 
+from packaging.tags import Tag
 from packaging.version import Version
-from pip._vendor.packaging.tags import Tag
 from pip._vendor.requests import Session
 
 from pdm._types import Distribution, Source
@@ -557,3 +557,13 @@ def deprecation_warning(
         if parsed_version >= Version(raise_since):
             raise DeprecationWarning(message)
     warnings.warn(message, DeprecationWarning, stacklevel=stacklevel + 1)
+
+
+def import_pip_vendor_object(module_name: str, attr: str) -> tuple:
+    """Import an attribute from both <module_name> and pip._vendor.<module_name>"""
+    from importlib import import_module
+
+    return (
+        getattr(import_module(module_name), attr),
+        getattr(import_module("pip._vendor." + module_name), attr),
+    )

